File: scripts/preview_gui.py
"""Development browser view of the shared Python game controller."""

from __future__ import annotations

import argparse
import contextlib
import json
import os
import queue
import secrets
import signal
import threading
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from pathlib import Path

import pygame

from pefforza.cli.gui_view import HEIGHT, TIER_COPY, WIDTH, cell_center
from pefforza.cli.lessons import LESSONS
from pefforza.cli.play_gui import GameApp
from pefforza.constants import COLS, ROWS


def parse_event(payload: dict) -> pygame.event.Event:
    """Accept only the input primitives used by the preview page."""
    kind = payload.get("kind")
    if kind == "select":
        column = payload.get("column")
        if type(column) is not int or not 0 <= column < COLS:
            raise ValueError("Invalid column")
        return pygame.event.Event(pygame.MOUSEMOTION, pos=cell_center(0, column))
    if kind in ("click", "move"):
        x, y = payload["x"], payload["y"]
        if type(x) is not int or type(y) is not int or not (0 <= x < WIDTH and 0 <= y < HEIGHT):
            raise ValueError("Invalid coordinates")
        event_type = pygame.MOUSEBUTTONDOWN if kind == "click" else pygame.MOUSEMOTION
        return pygame.event.Event(event_type, pos=(x, y), button=1)
    keys = {
        **{str(i): pygame.K_1 + i - 1 for i in range(1, 8)},
        "r": pygame.K_r,
        "u": pygame.K_u,
        "h": pygame.K_h,
        "d": pygame.K_d,
        "l": pygame.K_l,
        "n": pygame.K_n,
        "ArrowLeft": pygame.K_LEFT,
        "ArrowRight": pygame.K_RIGHT,
        "Enter": pygame.K_RETURN,
        " ": pygame.K_SPACE,
    }
    if kind == "key" and payload.get("key") in keys:
        return pygame.event.Event(pygame.KEYDOWN, key=keys[payload["key"]])
    raise ValueError("Unsupported input")


def snapshot(app: GameApp, now: int) -> dict:
    """One coherent state, including timing for browser-local interpolation.

    The board remains pre-drop until the controller commits the move. An
    animation's start tick is its identity; repeated snapshots must not restart it.
    """
    game = app.session
    lesson = LESSONS[app.lesson_index] if app.lesson_index is not None else None
    animation = None
    if app.animation is not None:
        column, row, player, start = app.animation
        animation = {
            "id": start,
            "column": column,
            "row": row,
            "player": player,
            "duration": app.animation_duration,
            "elapsed": max(0, now - start),
        }
    return {
        "rows": ROWS,
        "columns": COLS,
        "moves": len(game.moves),
        "board": game.board.tolist(),
        "player": game.current_player,
        "gameOver": game.game_over,
        "winner": game.winner,
        "winningCells": game.winning_cells,
        "difficulty": app.opponent_label,
        "opponentCopy": (
            TIER_COPY[app.difficulty][1]
            if app.worker.active_difficulty == app.difficulty
            else "Requested opponent unavailable. The active fallback is shown above."
        ),
        "lesson": app.lesson_index,
        "lessonTitle": lesson.title if lesson else None,
        "lessonCount": len(LESSONS),
        "lessonSolved": app.lesson_solved,
        "lessonAttempted": app.lesson_attempted,
        "selectedColumn": app.hint_col if app.hint_col is not None else app.selected_col,
        "status": app.notice or game.status,
        "animation": animation,
    }


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", type=int, default=3000)
    args = parser.parse_args()
    token = secrets.token_urlsafe(24)
    page = (
        Path(__file__)
        .with_name("preview_gui.html")
        .read_text(encoding="utf-8")
        .replace("__TOKEN__", token)
    )
    javascript = Path(__file__).with_name("preview_gui.mjs").read_bytes()
    inputs: queue.Queue[pygame.event.Event] = queue.Queue(maxsize=64)
    state = b"{}"

    class Handler(BaseHTTPRequestHandler):
        def respond(self, code: int, content: bytes, content_type: str) -> None:
            self.send_response(code)
            self.send_header("Content-Type", content_type)
            self.send_header("Content-Length", str(len(content)))
            self.send_header("Cache-Control", "no-store")
            self.send_header("X-Content-Type-Options", "nosniff")
            self.end_headers()
            with contextlib.suppress(ConnectionError):
                self.wfile.write(content)

        def do_GET(self) -> None:
            path = self.path.split("?", 1)[0]
            if path == "/":
                self.respond(200, page.encode(), "text/html; charset=utf-8")
            elif path == "/preview_gui.mjs":
                self.respond(200, javascript, "text/javascript; charset=utf-8")
            elif path == "/state":
                self.respond(200, state, "application/json")
            else:
                self.respond(404, b"Not found", "text/plain")

        def do_POST(self) -> None:
            if self.path != "/event" or self.headers.get("X-Preview-Token") != token:
                self.respond(403, b"Forbidden", "text/plain")
                return
            try:
                length = int(self.headers.get("Content-Length", "0"))
                if not 0 < length <= 1024:
                    raise ValueError("Invalid payload length")
                payload = json.loads(self.rfile.read(length))
                if not isinstance(payload, dict):
                    raise ValueError("Expected input object")
                inputs.put_nowait(parse_event(payload))
            except (ValueError, KeyError, TypeError, queue.Full):
                self.respond(400, b"Invalid input", "text/plain")
                return
            self.respond(204, b"", "text/plain")

        def log_message(self, format: str, *args: object) -> None:
            pass

    os.environ.setdefault("SDL_VIDEODRIVER", "dummy")
    os.environ.setdefault("SDL_AUDIODRIVER", "dummy")
    pygame.display.init()
    pygame.display.set_mode((1, 1))
    app = GameApp(seed=4)
    signal.signal(signal.SIGINT, signal.default_int_handler)
    signal.signal(signal.SIGTERM, lambda *_: setattr(app, "running", False))
    state = json.dumps(snapshot(app, pygame.time.get_ticks())).encode()
    server = ThreadingHTTPServer((args.host, args.port), Handler)
    server.daemon_threads = True
    threading.Thread(target=server.serve_forever, daemon=True).start()
    print(f"Pefforza browser preview on http://{args.host}:{args.port}", flush=True)
    clock = pygame.time.Clock()
    try:
        while app.running:
            now = pygame.time.get_ticks()
            for event in pygame.event.get():
                app.handle_event(event, now)
            for _ in range(64):
                try:
                    event = inputs.get_nowait()
                except queue.Empty:
                    break
                app.handle_event(event, now)
            app.tick(now)
            state = json.dumps(snapshot(app, now)).encode()
            clock.tick(60)
    except KeyboardInterrupt:
        pass
    finally:
        app.close()
        server.shutdown()
        server.server_close()
        pygame.quit()


if __name__ == "__main__":
    main()

File: scripts/preview_gui.mjs
// Python owns the game. This view only lays out state and interpolates a drop.
export class DropTimeline {
  constructor() { this.drop = null; this.startedAt = 0; }
  sync(drop, receivedAt, transitMs = 0) {
    if (!drop) { this.drop = null; return; }
    if (!this.drop || this.drop.id !== drop.id) {
      this.startedAt = receivedAt - drop.elapsed - Math.max(0, transitMs);
    }
    this.drop = drop;
  }
  progress(now, reducedMotion = false) {
    if (!this.drop) return null;
    if (reducedMotion) return 1;
    return Math.min(1, Math.max(0, (now - this.startedAt) / this.drop.duration));
  }
}

if (typeof document !== 'undefined') {
  const $ = id => document.getElementById(id);
  const token = document.querySelector('meta[name="preview-token"]').content;
  const game = $('game');
  const namespace = 'http://www.w3.org/2000/svg';
  const colors = {1:'#f98478', 2:'#efcd6f'};
  const timeline = new DropTimeline();
  const reducedMotion = matchMedia('(prefers-reduced-motion: reduce)');
  let state = null, connected = false, boardKey = '', animationFrame = null;
  let pieces, moving, highlight, columnButtons = [], lastHover = null;
  let commandQueue = Promise.resolve();

  function svg(tag, attributes, parent) {
    const element = document.createElementNS(namespace, tag);
    for (const [key, value] of Object.entries(attributes)) element.setAttribute(key, value);
    if (parent) parent.append(element);
    return element;
  }
  function text(id, value) {
    if ($(id).textContent !== value) $(id).textContent = value;
  }
  function piece(player, parent) {
    const group = svg('g', {}, parent);
    svg('circle', {r:29, fill:colors[player], stroke:'#111622', 'stroke-width':1}, group);
    svg('circle', {r:19, fill:'none', stroke:'#111622', 'stroke-opacity':.28, 'stroke-width':2}, group);
    const glyph = svg('text', {x:0, y:5, 'text-anchor':'middle', fill:'#111622', 'font-size':16}, group);
    glyph.textContent = player === 1 ? 'X' : 'O';
    return group;
  }
  function buildBoard(snapshot) {
    const width = snapshot.columns * 80, height = snapshot.rows * 80;
    game.setAttribute('viewBox', `-16 -40 ${width + 32} ${height + 60}`);
    svg('rect', {x:-16, y:-12, width:width+32, height:height+32, rx:28, fill:'#090d17'}, game);
    svg('rect', {x:-16, y:-16, width:width+32, height:height+32, rx:28, fill:'#263048', stroke:'#3d4a64'}, game);
    highlight = svg('rect', {y:0, width:70, height, rx:24, fill:'#2e3f52'}, game);
    for (let row=0; row<snapshot.rows; row++) {
      for (let col=0; col<snapshot.columns; col++) {
        svg('circle', {cx:col*80+40, cy:row*80+40, r:32, class:'slot'}, game);
      }
    }
    pieces = svg('g', {id:'pieces'}, game);
    moving = svg('g', {id:'moving-piece'}, game);
    const nav = document.querySelector('.column-controls');
    nav.style.gridTemplateColumns = `repeat(${snapshot.columns}, 1fr)`;
    for (let col=0; col<snapshot.columns; col++) {
      const button = document.createElement('button');
      button.textContent = col + 1;
      button.setAttribute('aria-label', `Drop in column ${col+1}`);
      button.addEventListener('click', () => send({kind:'key', key:String(col+1)}));
      nav.append(button);
      columnButtons.push(button);
    }
  }
  function canPlay() {
    return connected && state && state.player === 1 && !state.gameOver &&
      !state.animation && !state.lessonAttempted;
  }
  function select(column) {
    const active = canPlay() && state.board[0][column] === 0;
    highlight.setAttribute('visibility', active ? 'visible' : 'hidden');
    highlight.setAttribute('x', column*80+5);
    columnButtons.forEach((button, col) => {
      button.disabled = !canPlay() || state.board[0][col] !== 0;
      button.classList.toggle('selected', active && col === column);
    });
  }
  function drawDrop(now) {
    animationFrame = null;
    const progress = timeline.progress(now, reducedMotion.matches);
    if (progress === null) { moving.replaceChildren(); return; }
    const drop = timeline.drop;
    const y = -28 + ((drop.row*80+40) + 28) * progress**2;
    moving.setAttribute('transform', `translate(${drop.column*80+40} ${y})`);
    if (progress < 1) animationFrame = requestAnimationFrame(drawDrop);
  }
  function render(snapshot, receivedAt, transitMs) {
    state = snapshot;
    if (!pieces) buildBoard(state);
    const key = JSON.stringify([state.board, state.winningCells]);
    if (key !== boardKey) {
      pieces.replaceChildren();
      for (let row=0; row<state.rows; row++) {
        for (let col=0; col<state.columns; col++) {
          if (state.board[row][col]) {
            const token = piece(state.board[row][col], pieces);
            token.setAttribute('transform', `translate(${col*80+40} ${row*80+40})`);
          }
        }
      }
      for (const [row,col] of state.winningCells) {
        svg('circle', {cx:col*80+40, cy:row*80+40, r:35, class:'winning'}, pieces);
      }
      text('board', state.board.map(row => row.map(p => p===1?'X':p===2?'O':'.').join(' ')).join('\n'));
      boardKey = key;
    }
    const previousDrop = timeline.drop;
    timeline.sync(state.animation, receivedAt, transitMs);
    if (!state.animation || !previousDrop || previousDrop.id !== state.animation.id) {
      moving.replaceChildren();
      if (state.animation) piece(state.animation.player, moving);
    }
    if (animationFrame === null) drawDrop(performance.now());
    select(state.selectedColumn);
    const lesson = state.lesson !== null;
    text('headline', lesson ? 'Tiny tactics.' : 'Four in a row.');
    text('subtitle', lesson ? state.lessonTitle : 'One more round?');
    text('counter', lesson ? `Lesson ${state.lesson+1} / ${state.lessonCount}` : `Move ${String(state.moves + Number(!state.gameOver)).padStart(2,'0')}`);
    let turn = state.player === 1 ? 'Your turn' : 'Thinking…';
    if (state.gameOver) turn = state.winner===1 ? 'You got four!' : state.winner===2 ? 'AI got four.' : 'A worthy draw.';
    if (lesson) turn = state.lessonSolved ? 'You found it!' : state.lessonAttempted ? 'Try again' : 'Your challenge';
    if (state.animation) turn = state.animation.player===1 ? 'Nice drop.' : "AI’s move";
    text('turn', turn);
    text('status', state.status);
    $('status').dataset.moves = state.moves;
    $('status').dataset.gameOver = state.gameOver;
    text('opponent-heading', lesson ? 'Keep exploring' : 'Your opponent');
    text('difficulty', lesson ? 'Next lesson' : state.difficulty);
    text('difficulty-key', lesson ? 'N / next' : 'D / new round');
    text('opponent-copy', lesson ? 'One move, one idea. Retry freely; hints explain the answer.' : state.opponentCopy);
    text('restart-label', lesson ? 'Try again' : state.gameOver ? 'Play again' : 'New round');
    text('lessons-label', lesson ? 'Free play' : 'Tactical lessons');
    document.querySelector('[data-key="u"]').disabled = !lesson && state.moves===0 && !state.animation;
    document.querySelector('[data-key="h"]').disabled = Boolean(state.animation) || (!lesson && (state.gameOver || state.player!==1));
  }
  function disconnect() {
    connected = false;
    timeline.sync(null, performance.now());
    if (moving) moving.replaceChildren();
    if (state) select(state.selectedColumn);
    document.querySelectorAll('[data-key]').forEach(button => { button.disabled = true; });
    text('connection', 'Disconnected · reconnecting…');
  }
  function send(payload) {
    if (!connected) return;
    // Serialize controls so fast Undo / Restart cannot overtake an earlier drop.
    commandQueue = commandQueue.then(async () => {
      const response = await fetch('/event', {
        method:'POST', headers:{'Content-Type':'application/json','X-Preview-Token':token},
        body:JSON.stringify(payload), signal:AbortSignal.timeout(3000)
      });
      if (response.status === 403) {
        // A restarted local server creates a fresh session token.
        location.reload();
        return;
      }
      if (!response.ok) throw new Error('Input rejected');
    }).catch(disconnect);
  }
  document.querySelectorAll('[data-key]').forEach(button => {
    button.addEventListener('click', () => send({kind:'key',key:button.dataset.key}));
  });
  document.addEventListener('keydown', event => {
    if (event.target.closest('button') && ['Enter',' '].includes(event.key)) return;
    const key = event.key.length===1 ? event.key.toLowerCase() : event.key;
    if (['1','2','3','4','5','6','7','r','u','h','d','l','n','ArrowLeft','ArrowRight','Enter',' '].includes(key)) {
      event.preventDefault();
      if (!event.repeat) send({kind:'key',key});
    }
  });
  function pointerColumn(event) {
    if (!state) return null;
    const point = new DOMPoint(event.clientX,event.clientY).matrixTransform(game.getScreenCTM().inverse());
    const col = Math.floor(point.x/80);
    return col>=0 && col<state.columns && point.y>=-40 && point.y<state.rows*80 ? col : null;
  }
  game.addEventListener('click', event => {
    const column = pointerColumn(event);
    if (column !== null && canPlay()) send({kind:'key',key:String(column+1)});
  });
  game.addEventListener('pointermove', event => {
    const column = pointerColumn(event);
    if (column !== null && canPlay() && column !== lastHover) {
      lastHover = column;
      select(column);
      send({kind:'select',column});
    }
  });
  game.addEventListener('pointerleave', () => { lastHover = null; });
  reducedMotion.addEventListener('change', () => {
    if (timeline.drop && animationFrame === null) drawDrop(performance.now());
  });
  async function refresh() {
    const sentAt = performance.now();
    try {
      const response = await fetch('/state', {cache:'no-store',signal:AbortSignal.timeout(3000)});
      if (!response.ok) throw new Error('Preview unavailable');
      const snapshot = await response.json();
      const receivedAt = performance.now();
      connected = true;
      document.querySelectorAll('[data-key]').forEach(button => { button.disabled = false; });
      render(snapshot, receivedAt, (receivedAt-sentAt)/2);
      text('connection', 'Local game · one shared round');
    } catch { disconnect(); }
    // Network cadence does not set animation cadence: RAF draws between updates.
    setTimeout(refresh, document.hidden ? 250 : connected ? 32 : 500);
  }
  refresh();
}
